pkg/auth/remoteauthenticator/authenticator.go

```diff
@@ -0,0 +1,169 @@
+package remoteauthenticator
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/treeverse/lakefs/pkg/auth"
+	"github.com/treeverse/lakefs/pkg/auth/model"
+	"github.com/treeverse/lakefs/pkg/config"
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+const RemoteAuthSource = "remote_authenticator"
+
+// AuthenticationRequest is the request object that will be sent to the remote authenticator service as JSON payload in a POST request
+type AuthenticationRequest struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+// AuthenticationResponse is the expected response from the remote authenticator service
+type AuthenticationResponse struct {
+	// ExternalUserIdentifier is optional, if returned then the user will be used as the official username in lakeFS
+	ExternalUserIdentifier *string `json:"external_user_identifier,omitempty"`
+}
+
+// RemoteAuthenticator client
+type RemoteAuthenticator struct {
+	AuthService auth.Service
+	Logger      logging.Logger
+	Config      *config.RemoteAuthenticator
+	serviceURL  string
+	httpClient  *http.Client
+}
+
+func NewRemoteAuthenticator(conf *config.RemoteAuthenticator, authService auth.Service, logger logging.Logger) (auth.Authenticator, error) {
+	serviceURL, err := url.JoinPath(conf.BaseURL, conf.AuthEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	logger.WithField("auth_url", serviceURL).Info("initializing remote authenticator")
+
+	return &RemoteAuthenticator{
+		Logger:      logger,
+		Config:      conf,
+		serviceURL:  serviceURL,
+		AuthService: authService,
+	}, nil
+}
+
+func (ra *RemoteAuthenticator) client() *http.Client {
+	if ra.httpClient == nil {
+		ra.httpClient = http.DefaultClient
+		ra.httpClient.Timeout = ra.Config.RequestTimeout
+	}
```

Comment:
**The reason I used a function:** because for decoupling the instance from the caller for future use-cases (I admit it doesn't add any meaningful plus right now). 
You're right there is  a **race condition first time it's called**, I didn't want to add locking techniques  or something of that sort because there's no real harm if that occurs. 
Regarding the initiation inside the function and not in the constructor, I wanted to be **lazy about memory allocation**. 

Having said that: WDYT?